### ISSUE TITLE
A not-yet logged in IDP has no route to success with this flow

### ISSUE THREAD

**Issue by bvandersloot-mozilla** [fedidcg/FedCM#442]
Enabling an option in which the IDP can provide a way to be logged in (via a controlled call to window.open) and then immediately be dropped back into the account chooser would be nice! I discussed something related to this in https://github.com/fedidcg/proposals/issues/3 but it could be adapted here. I'll add a comment with a general shape of the proposal.

**Comment by cbiesinger** [comment]
Yes, this will be needed, especially if an RP wants to use FedCM for a button flow (as opposed to triggering it onload, where silent failure is probably acceptable)

#436 adds a signin_url field, which is part of what is needed here.

**Comment by bvandersloot-mozilla** [comment]
> #436 adds a signin_url field, which is part of what is needed here.

Agreed. A key piece will be having the proposed dialog that shows that link show up when a user isn't yet logged in. In addition to that, there are two options:

1. require the IDP support something like a solution to #441 so it can complete the flow in the opened link
2. Have the sign-in status API signal to pending login requests that have already "chosen" that IDP that they should re-issue the account-list request. If that fails it should display a failure popup.

I think the best way forward is the latter.

**Comment by cbiesinger** [comment]
Yes option 2 sounds good to me.

But I think we need to figure out how to trigger this behavior. Maybe add a flag somewhere saying that if there are no accounts found, that we should show a signin prompt instead of returning failure silently?

And/or have the presence of transient user activation trigger this behavior?

**Comment by asr-enid** [comment]
> But I think we need to figure out how to trigger this behavior. Maybe add a flag somewhere saying that if there are no accounts found, that we should show a signin prompt instead of returning failure silently?

I think this is also another instance where the RP would want a flag to influence this behavior especially with the introduction of multi-idp support - see https://github.com/fedidcg/FedCM/pull/438#discussion_r1110156822 - there could be a IDP specific flag where the RP could indicate to be wanting to enable a IDP sign-in from its site.

**Comment by Bojhan** [comment]
This would be quite relevant to Seamless Access. We expect that users will move between many sites, wanting to access for example scientific content.

They will be shown the option, but it needs to be clear that they are signed in or still need to

**Comment by samuelgoto** [comment]
> Enabling an option in which the IDP can provide a way to be logged in (via a controlled call to window.open) and then immediately be dropped back into the account chooser would be nice! I 
> > https://github.com/fedidcg/FedCM/pull/436 adds a signin_url field, which is part of what is needed here.

> Agreed.

+1 that the `signin_url` can be the url that we point to in the window as @bvandersloot-mozilla described above.

> But I think we need to figure out how to trigger this behavior. Maybe add a flag somewhere saying that if there are no accounts found, that we should show a signin prompt instead of returning failure silently?

Yeah, maybe an extra attribute in the [`configURL`](https://fedidcg.github.io/FedCM/#idp-api-config-file) like `on_signedout: signin || dismiss` (we currently `dismiss`, so maybe that can be the default so that we can keep backwards compatibility?)? 

The triggering seems like something that IdPs should have the control over rather than RPs/callers, so my first intuition is that it should be in the [config file](https://fedidcg.github.io/FedCM/#idp-api-config-file). 

```json
{
  "on_signedout": "signin",
  "signin_url": "https://idp.example/login",
}
```

And then have the [create identity credential algorithm](https://fedidcg.github.io/FedCM/#create-identity-credential) open a window with the `signin_url` when the user is not logged in to the IdP?

@bvandersloot-mozilla does this sound more or less right to you?

**Comment by cbiesinger** [comment]
hm, I think RPs may want different behavior at different times. For example, they may want to show an account chooser onload but only if the user is logged in; but they may have a "sign in" button which should not fail silently.

**Comment by samuelgoto** [comment]
> hm, I think RPs may want different behavior at different times. For example, they may want to show an account chooser onload but only if the user is logged in; but they may have a "sign in" button which should not fail silently.

Yeah, I was thinking that IdPs would have the ability to have more than one `configURL` that they can expose for different products that behave one way (e.g. account choosers onload) or another (button flows).

So, RPs can still choose between two different `configURLs`, but IdPs also have the controls to expose or not one/both/either option.

**Comment by bvandersloot-mozilla** [comment]
I think that enumerating a set of supported flows may be painting ourselves into a corner. 

How does either proposal interact with a multiple-IDP use case?

**Comment by philsmart** [comment]
I've tried this with the multi-idp mode. As it is a global property, it seems to pop up the `signin_url` window for every IdP that you've not signed into before offering the account chooser. I am not sure how sensible that is.

**Comment by npm1** [comment]
If that comment is about the Chrome flag, we haven't worked on that in a while so it is likely not working correctly at this point. In multi-idp mode, I'd expect the user to pick the IDP they want to sign in to, not for one popup per IDP to open.

**Comment by philsmart** [comment]
That's it (chrome flag), thanks. The behavior I observed was a popup per signed-out IdP. But I realize the multi-idp support is a WiP, so just adding a data point from testing.

**Comment by asr-enid** [comment]
> The behavior I observed was a popup per signed-out IdP. But I realize the multi-idp support is a WiP, so just adding a data point from testing.

Interesting, thats really not what you'd want for a final version. As noted in the CG call I think the RP/Authentication Intermediary should be able top open a pop-up in this case to let the user choose how to proceed (not a browser mediated UI, or at least both options must be available)

See also the implementation bug in Chrome to track this: https://bugs.chromium.org/p/chromium/issues/detail?id=1487268

**Comment by asr-enid** [comment]
@samuelgoto seems this should be a proposal under https://github.com/fedidcg/FedCM/tree/main/proposals ?

**Comment by samuelgoto** [comment]
Just a small update on this: we built a prototype in chrome canaries with the following formulation:

```javascript
await navigator.credentials.get({
  identity: {
    mode: "button" || "widget"
    providers: [{
      // ...
    }]
  }
});
```

You can try this by flipping this flag in chrome canaries: `chrome://flags/#fedcm-authz`.

What it does is that, on `mode=button`, when the user is `logged-out` (based on what the browser knows about the Login Status API), it opens a pop-up window pointing the user to the `login_url` found in the `configURL` passed.

Any chance we have anyone around interested in giving this a try and letting us know if this behaves the way that they'd expect this to behave?

**Comment by asr-enid** [comment]
> Any chance we have anyone around interested in giving this a try and letting us know if this behaves the way that they'd expect this to behave?

FWIW - this is also already supported as an option in the demonstrator implementations (mode = button) on the RP side.

**Comment by philsmart** [comment]
I should add, that it behaves in the way described for a single IdP.

**Comment by philsmart** [comment]
I have no good suggestions of how to fix that for multi-idps, even if you moved the mode option into the provider's array (i.e., per-provider) I am not sure how the RP would know to set it differently per IdP (or even why it would do that), and would probably just set it the same for each provider anyway (so I think it makes sense to set globally as it is). I guess you might need the user to select which IdP to use first before trying to retrieve accounts from it, and therefore before deciding if the sign-in popup is needed.

**Comment by cbiesinger** [comment]
Yeah, we'd probably have to have an "IDP selection" dialog if the user is not signed in to any of the IDPs, and if the user is signed in to at least one IDP the dialog should still let the user sign in to one of the others

**Comment by hlflanagan** [comment]
Don't forget that SeamlessAccess has a lot of experience with that IdP choosing process. Lessons to be learned there (and possibly code to use).

**Comment by asr-enid** [comment]
> So the account selector would never need to show accounts for more than one IdP (which it currently allows).

Could well be that the user is signed into multiple not just one - so this assumption might not work out. 

> Yeah, we'd probably have to have an "IDP selection" dialog if the user is not signed in to any of the IDPs, and if the user is signed in to at least one IDP the dialog should still let the user sign in to one of the others

Agree - there must be a way to use an IDP to which the user is currently not signed into, in cases he is with at least one other

**Comment by cbiesinger** [comment]
Hmm @asr-enid may want a single button that works for GMX/web.de/etc?

In general I would like to think that we can be more ambitious than the current state of the web and it would be nice to have a single "Log in" button that supports all the IDPs

**Comment by philsmart** [comment]
Yes, you could probably do the IdP discovery part external to the FedCM API—whether that is a list of buttons (sign in with X.com) or something more sophisticated when there are thousands of possible IdPs—and then seed FedCM with that choice. I guess in which case you would never directly need to support more than one `provider` in FedCM? Unless that is still useful in the `widget` mode, I am not sure.

I was thinking it would operate more along the lines of what @cbiesinger said, a single button that ran through the entire selection flow. This does become hard in the 'hundreds or thousands of IdPs R&E world', however.

**Comment by npm1** [comment]
I think supporting multiple IdPs in a button flow will allow RPs to provide a better user experience. Providing this means that the user would not need to remember what account they had, which IdP they used before, or even if they already had an account or not.

**Comment by asr-enid** [comment]
> Hmm @asr-enid may want a single button that works for GMX/web.de/etc?

Absolutely

**Comment by yi-gu** [comment]
We have some [mocks](https://docs.google.com/presentation/d/1iURrPakaHgBfQ6mAefKijjxToiTTgBSPz1rtaV0od98/edit?usp=sharing) to show how we plan to support the button flow in Chrome. We can go over the CUJs in the next CG call but in the meantime please feel free to leave your feedback here.

**Comment by philsmart** [comment]
I think (not always easy to see if I am doing something wrong, or the API) in button mode, the browser is sending different cookies in the popup window (signin_url) than those it sends as part of the FedCM accounts/assertion call. I'll keep digging, but maybe there is an easy official answer.

**Comment by yi-gu** [comment]
Summarizing the discussions above to move forward:

# Context
The button flow differs from the widget flow in several ways. The most significant difference is that the button flow is on the critical path of a user's sign-in journey. This means that a user must be able to successfully sign in with a federated account using this flow. In contrast, the widget flow is an optimized flow that can reduce sign-in friction. This means that if the widget flow is unavailable, a user can still click a "Sign in with IdP" button to continue.

In view of the complexity of the critical sign-in flow, as well as the diversity of deployment environments (e.g., an RP may or may not use an IdP SDK), it is necessary to explicitly define the responsibilities of all parties involved.

- What IdP controls
  - Whether a user can sign in to the IdP if they don’t have an active IdP session
  - Whether a user can sign in with new IdP accounts even if they have active IdP sessions 
- What RP controls
  - Note: we consider something RP controlled if it’s defined on the RP site even if an IdP can do it directly with SDK.
  - Whether to use the FedCM button flow and/or widget flow when user tries to sign in
- What browser controls
  - Whether to gate the button flow behind a user gesture
  - Whether to make the button flow more prominent such as controlling the UI’s modality and/or position
  - How to handle the coexistence of both flows
  - Whether to let IdP know which flow was triggered in the ID assertion endpoint

# Proposal
## IdP control
```
// IdP's config file
{
  "accounts_endpoint" : ...,
  "modes: {
    "button": {
      "supports_use_other_account": true|false,
    },
    "widget": {
      "supports_use_other_account": true|false,
    }
  }
}
```
## RP control
```
await navigator.credentials.get({
  identity: {
    providers: [
      {
        configURL: "https://idp.example/fedcm.json",
        clientId: "client1234",      
      },
    ],
    mode: "button"|"widget",
  },
});
```
## Browser control
Include `mode` in the POST request such that IdP can know which flow the user has gone through.

POST /fedcm_assertion_endpoint HTTP/1.1
Host: idp.example
Origin: https://rp.example/
Content-Type: application/x-www-form-urlencoded
Cookie: 0x23223
Sec-Fetch-Dest: webidentity

account_id=123&client_id=client1234&nonce=Ct0D&disclosure_text_shown=true&is_auto_selected=false&**mode=button**
## Overview
<p align="center">
     <img width="467" src="https://github.com/yi-gu/fedcm-images/blob/main/button_flow/ButtonFlow.jpeg?raw=true"/>
</p>

# Open questions
## Mode options
The current option "button" and "widget" are not ideal because they are focused on the UI affordance instead of the essential distinction of the two types of flows. What "button flow" implies is that it's user initiated and the user must be served by the UA to complete the sign-in flow. This is a critical user journey so the browser should mediate it in a prominent way. What "widget flow" implies is that it may not be gated by user activation therefore it should be treated as an optimized sign-in option and it's OK to fail silently. In other words, it's up to the UA how to implement these two flows with different UI affordances. e.g. one may implement the optional flow with quieter UI like a chip on the URL bar.

We will try to gather more feedback on the naming before shipping.

## Feature detection
There's no dedicated way to feature detect dictionary members. There have been [discussions](https://github.com/whatwg/webidl/issues/107) over the years but folks haven't been sufficiently aligned on the proposals. In the interim,  developers can use the following way for feature detection for better interoperability.
```
var supportsFedCmMode = false;
try {
   navigator.credentials.get({
       identity:
     Object.defineProperty(
     {}, 'mode', {
       get: function () { supportsFedCmMode = true; }
     }
   )
   });
} catch(e) {}
```

We can revisit this decision if the community has other thoughts or better options.

# Security Considerations
The button mode shares most of the security properties from the [widget mode](https://fedidcg.github.io/FedCM/#security). e.g. honoring CSP, CORS, using security headers, not asking users to type in the browser UI etc.. It’s worth noting that the pop-up window has the same web platform properties as what one would get with window.open(url,””,”popup,noopener,noreferrer”)) that loads the login_url. There's no communication between the website and this pop-up is allowed (e.g. no postMessage, no window.opener).

# Privacy Considerations
The button mode shares all the privacy properties from the [widget mode](https://fedidcg.github.io/FedCM/#privacy). There are some notable differences:
- The button mode should be gated by user activation.
- For the widget flow which is not gated by user gesture, we do not send requests to IdP if the IdP’s login status is “logged-out”. We propose to do the same for the button flow. However, we should trigger the sign in to IdP flow instead of failing silently because this is on the critical sign-in path.
- The mismatch UI from LoginStatus API is replaced by a louder pop-up window in cases where a silent timing attack could happen.
- The new sign in to IdP affordance has a higher privacy bar because it does not append any parameters to the `login_url`

**Comment by timcappalli** [comment]
@yi-gu re: feature detection, we could consider something like [getClientCapabilities()](https://w3c.github.io/webauthn/#sctn-getClientCapabilities) which we recently added to WebAuthn L3.

**Comment by yi-gu** [comment]
Thanks @timcappalli for the suggestion! That idea was brought up late 2023 when it was un-merged. Will take a closer look and we can discuss about it in the CG.

**Comment by yi-gu** [comment]
Plan to share some update from the Chrome team on this topic in the next CG call. In addition, we could discuss about Tim's suggestion on feature detection.

**Comment by hlflanagan** [comment]
Question for @bvandersloot-mozilla - is the button mode as currently being tested in origin trials sufficient to resolve this issue?

**Comment by RByers** [comment]
For context, [here](https://developers.google.com/privacy-sandbox/blog/fedcm-chrome-125-updates) are some details on Chrome's FedCM button flow origin trial.

**Comment by bvandersloot-mozilla** [comment]
Yeah, the button flow is sufficient for our concern. We agree that multiple cold IDPs is a much harder problem, and may not be desirable for the browser to solve.

**Comment by yi-gu** [comment]
We added some [mocks](https://docs.google.com/presentation/d/1iURrPakaHgBfQ6mAefKijjxToiTTgBSPz1rtaV0od98/edit#slide=id.g2e9c3a70a9b_0_26) for Android. They are not final and we will keep folks updated as we are getting closer to running Origin Trials in Chrome.

**Comment by garykac** [comment]
There's not a lot of information in the Explainer post about the proposed Use Other Account API/extension here.

Is it using an existing account chooser? If so, why is it called out as a new proposal?

If it is different from (or a modification to) an existing account chooser, then there needs to be more info in order to perform security and privacy reviews. How is it different and how does it ensure that available accounts and current logged-in status is not leaked?

**Comment by yi-gu** [comment]
Hi garykac@,
From API's perspective, "Use Other Account" introduces a new way for IdPs to express that whether they allow using different account sessions and that's it. From privacy and security's perspective, the "log in to IdP" flow reuses the existing pattern from the LoginStatus API.
From UI's perspective, this feature may be presented differently to users across user agents. In Chrome's proposal, we add a  "Use a different account" button to the account chooser (see mocks [here](https://docs.google.com/presentation/d/1iURrPakaHgBfQ6mAefKijjxToiTTgBSPz1rtaV0od98/edit#slide=id.g2b23fb91b23_0_85)). Upon user clicking that button, we trigger the same UI affordance for users to sign in to the IdP with the `login_url` from the LoginStatus API.

**Comment by garykac** [comment]
OK, thank you for the clarification. I hoped that was the case, but ChromeStatus describes this as a proposed API, which makes it seem like a larger change than it actually is.

**Comment by yi-gu** [comment]
Sorry for the confusion. It is indeed a new API like
```
// IdP's config file
{
  "accounts_endpoint" : ...,
  "modes: {
    "button": {
      "supports_use_other_account": true|false,
    },
    "widget": {
      "supports_use_other_account": true|false,
    }
  }
}
```
but as you mentioned it was not a big API change with new privacy/security implications.
